win_scheduled_task, added hourly and interval options, update bug fix

DepthDeluxe opened the pull request:
- frequency can be set to "hourly" and "interval"
- hourly will execute once an hour
- interval lets user specify an interval between executions, user can
specify units of interval
- powershell script had some code that couldn't get executed, if ever got activated would have been missing the check to see if trigger changed

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
win_scheduled_task

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.2.0.0
  config file =
  configured module search path = ['/Users/cheinzma/ansible-library', '/Users/cheinzma/Documents/ansible/lib/modules']
```

##### SUMMARY
Adding interval ability for windows scheduled tasks.  Added "hourly" and "interval" options for frequency, "interval" optional argument and "interval_unit" optional argument.

The "hourly" frequency option allows a task to be executed once an hour.  The time argument specifies when it should run, the next executions will happen once an hour.

The "interval" frequency option allows a task to be executed at a user-supplied interval.  The "interval" argument is the amount of time between executions, and the "interval_unit" is the units that number is in.  Supported units are "seconds", "minutes", "hours", and "days".  The powershell script checks to make sure that "interval_unit" is set to those four.

fixes issue #20056

DepthDeluxe committed:
win_scheduled_task, added hourly and interval options, update bug fix
- frequency can be set to "hourly" and "interval"
- hourly will execute once an hour
- interval lets user specify an interval between executions, user can
specify units of interval
- powershell script had bug where task wouldn't update if just the
trigger component changed, since option

addressed PR issues, added check mode

if task doesn't exist on Windows Server 2012R2, exception gets thrown

DepthDeluxe commented on lib/ansible/modules/windows/win_scheduled_task.ps1:
```diff
@@ -65,7 +79,7 @@ else
 
 try {
-    $task = Get-ScheduledTask -TaskPath "$path" | Where-Object {$_.TaskName -eq "$name"}
+    $task = Get-ScheduledTask -TaskPath "$path" -TaskName "$name"
```
This is a much cleaner way to achieve the same thing

DepthDeluxe commented on lib/ansible/modules/windows/win_scheduled_task.ps1:
```diff
@@ -144,21 +164,15 @@ try {
         $result.changed = $true
     }
     elseif( ($state -eq "present") -and ($exists -eq $true) ) {
-        if ($task.Description -eq $description -and $task.TaskName -eq $name -and $task.TaskPath -eq $path -and $task.Actions.Execute -eq $execute -and $taskState -eq $enabled -and $task.Principal.UserId -eq $user) {
```
As far as I could tell the `true` codepath never gets executed even though task details don't change.  In addition, I realized this never checks to see if the trigger ever changes.

dagwieers commented on lib/ansible/modules/windows/win_scheduled_task.ps1:
```diff
@@ -25,6 +25,8 @@ $ErrorActionPreference = "Stop"
+$interval_unit = Get-Attr $params -name "interval_unit" -default minutes
 $enabled = Get-AnsibleParam $params -name "enabled" -default $true
 $enabled = $enabled | ConvertTo-Bool
 $description = Get-AnsibleParam $params -name "description" -default " "
```
So nowadays we follow this syntax:

```powershell
$enabled = Get-AnsibleParam -obj $params -name "enabled" -type "bool" -default $true
$description = Get-AnsibleParam -obj $params -name "description" -type "str" -default " "
...
```

Possible types currently are:
- str
- path
- bool
- int

Also, I wonder why description defaults to " ", that seems a bit weird.

DepthDeluxe replied:
Will fix `Get-Attr` -> `Get-AnsibleParam`.

Description probably defaults to "" because it is unnecessary to make it required and original dev's didn't want to add support for checking if description field exists.  Personally, I think it's better to leave it as is since it honestly doesn't matter what the description is.

dagwieers commented on lib/ansible/modules/windows/win_scheduled_task.ps1:
```diff
@@ -25,6 +25,8 @@ $ErrorActionPreference = "Stop"
 $params = Parse-Args $args;
```
The trailing semi-colon is not needed, but here it would be nice if we could add check-mode support. See the win_shortcut plugin for the exact syntax to add this to your module.

In check-mode we would like the plugin to follow as much code-paths or failure mode as possible without making an effective change. So some modules include the connection/disconnections as part of the check-mode logic, or all possible pre-flight checks a module has.

We want the module to fail exactly the same way it would fail in real run-mode. Except without making the change. In some cases we simply add the -WhatIf option to the command to let Powershell do as much as possible without the actual change.

dagwieers commented on lib/ansible/modules/windows/win_scheduled_task.ps1:
```diff
@@ -54,6 +56,18 @@ if ($frequency -eq "weekly")
 }
+elseif ($frequency -eq "interval")
+{
+    if (!($interval))
```
I personally prefer this:
```powershell
    if (-not $interval) {
        Fail-Json $result "missing required argument: interval"
    }
```
```

dagwieers commented on lib/ansible/modules/windows/win_scheduled_task.ps1:
```diff
@@ -25,6 +25,8 @@ $ErrorActionPreference = "Stop"
 
 $days_of_week = Get-AnsibleParam $params -name "days_of_week"
+$interval = Get-Attr $params -name "interval" -default $null
+$interval_unit = Get-Attr $params -name "interval_unit" -default minutes
```
interval unit can only be one out of a list. You can do this:
```powershell
$interval_unit = Get-AnsibleParam -obj $params -name "interval_unit" -type "bool" -default "minutes" -validateset "days","hours","minutes","seconds"
```

And then you don't have to check below anymore if it is one of these options.

dagwieers commented on lib/ansible/modules/windows/win_scheduled_task.ps1:
```diff
@@ -99,14 +113,20 @@ try {
     Set-Attr $result "exists" "$exists"
```
If $result becomes a dictionary rather than a PSObject above;
```powershell
 $result = @{
    changed = $false
}
```
you can simply do:
```powershell
$result.exists = $exists
```
eveywhere in the module.

dagwieers commented:
@DepthDeluxe Good work ! There are plenty of small things to change in the code that do not originate from you. But Windows module development is evolving rapidly so we should take advantage of the new possibilities :-)

Check-mode is a must these days for Windows modules, if you could add that as well that would be great !

DepthDeluxe commented:
@dagwieers, addressed your concerns and added `--check` mode support.  Validated that it does work.  Prevents schedule changes and prevents deletion of tasks that currently exist.

I appreciate the feedback, still learning the wonders of powershell 😄

DepthDeluxe commented:
Also added another module `win_proxy`, allows users to configure the web proxy settings for a particular user.

Validated that `--check` mode works, that it properly sets the configuration, and is idempotent.

mattclay commented:
Closing and re-opening to trigger CI.

dagwieers commented:
@DepthDeluxe Please open a separate PR for win_proxy !

DepthDeluxe commented:
@dagwieers, sorry about that.  Purged this module from the PR, kept clean git history.

You have a good point about the module conventions.  I figured that it would be nice in terms of ease of use for the end-user.  I'm still trying to get a good feeling of what should be in a module vs. role with input arguments.

dagwieers commented:
@DepthDeluxe No worries, we are entering uncharted territory here. With the upside that we can plant new flags and build new cities as we go along !

dagwieers commented on lib/ansible/modules/windows/win_scheduled_task.ps1:
```diff
@@ -22,17 +23,21 @@ $ErrorActionPreference = "Stop"
+$interval = Get-AnsibleParam -obj $params -name "interval" -type "int" -default $null
+$interval_unit = Get-AnsibleParam -obj $params -name "interval_unit" -type "str" -default "minutes"
 $enabled = Get-AnsibleParam $params -name "enabled" -default $true
 $enabled = $enabled | ConvertTo-Bool
```
Please make that:
```powershell
$enabled = Get-AnsibleParam -obj $params -name "enabled" -type "bool" -default $true
```

PS I also noticed that some of my older comments (like this one) have not been adapted :-/

jhawkesworth commented:
Would be nice to have some integration tests, but the changes look good to me.

shipit

jhawkesworth commented:
Actually I've re-read @dagwieers comments above and I think having modules to wrap registry changes isn't the way forward.  
I think roles containing template and win_regmerge/win_regedit tasks are the way this should be done.
Would be good to make a sort of 'template role' where you could slot in your registry templates, but some registry tweaks need a whole set of changes (best suited to win_regmerge) and others are single registry entries (better suited to win_regedit), so a single 'template role' probably wouldn't be enough.
Will think about this further.

DepthDeluxe commented on lib/ansible/modules/windows/win_scheduled_task.ps1:
```diff
@@ -42,18 +43,26 @@ $state = Get-AnsibleParam -obj $params -name "state" -type "str" -default "prese
 try {
-    $task = Get-ScheduledTask -TaskPath "$path" | Where-Object {$_.TaskName -eq "$name"}
+    try {
+        $task = Get-ScheduledTask -TaskPath "$path" -TaskName "$name"
```
This needs to be wrapped in a try statement because the script fails when the task doesn't exist, at least on Windows Server 2012R2.

DepthDeluxe commented:
@dagwieers finally found the time to fix the issues and catch up with the 2.3 update.  I have validated that this works in Windows Server 2012R2.

jborean93 commented:
concur with @jhawkesworth about the tests but this PR has been sitting here for a while and would be good to get this feature in.

shipit

jborean93 commented:
bot_status

jhawkesworth commented:
@DepthDeluxe - please could you rebase this pull request?  Sorry there are now conflicts to resolve following merge of fix for taskpaths.

DepthDeluxe committed:
added path sanitization, proper unregistering of scheduled tasks

DepthDeluxe commented on lib/ansible/modules/windows/win_scheduled_task.ps1:
```diff
@@ -135,19 +152,25 @@ try {
 
     if ( ($state -eq "absent") -and ($exists) ) {
-        Unregister-ScheduledTask -TaskName $name -Confirm:$false -WhatIf:$check_mode
+        Unregister-ScheduledTask -TaskName $name -TaskPath $path -Confirm:$false -WhatIf:$check_mode
```
We should put in the -TaskPath flag here to guarantee that we are removing the task from the desired folder.

DepthDeluxe commented on lib/ansible/modules/windows/win_scheduled_task.ps1:
```diff
@@ -77,25 +78,41 @@ $description = Get-AnsibleParam -obj $params -name "description" -type "str" -de
 
+# Ensure the TaskPath starts with and ends with slashes as this causes issues with some
+# versions of Windows
+if ($path[0] -ne '\') {
```
This newly added sanitization will try to fix invalid TaskPath syntax that may cause confusing errors for the user.

For example if a user specifies

```
    - name: Backup Task
      win_scheduled_task:
        name: Backup
        execute: E:\TableauAdmin\scripts\ServerBackup.ps1
        frequency: daily
        time: 8:00pm
        user: SYSTEM
        state: present
        path: Tableau
```

The current script will throw an error `Cannot create a file that already exists`.  This is because the function `Unregister-ScheduledTask` throws an exception that it can't find the given task.  This is due to the fact that `$path = Tableau` when it is really expecting the path to both start with and end with a `\`.

The sanitization guarantees that even if a user forgets to have the path start with and end with `\`, the script still runs as expected as now `$path = \Tableau\` in this instance.

dagwieers replied:
I would rather fail and return a message back to the user. Because we cannot assume what the user wants. Adding a trailing slash is no problem, but prepending is not a good idea IMO

DepthDeluxe replied:
@dagwieers except that how would it ever be possible to get anything but the root level?  Relative addressing requires something to be relative to and there is no possible option to set in this case.

dagwieers replied:
@DepthDeluxe Not sure what you are getting at. What I am saying is that I want to force the user to provide an absolute path only. If it does not start with a slash, it should fail with an error. Not pre-pend a slash ourselves.

DepthDeluxe replied:
@dagwieers understood, I'll make the change.

I was trying to explain that I'm trying to apply the "be liberal in what you accept" philosophy.  In this case there isn't really ambiguity between having a slash and missing a slash.  Missing the slash at the beginning currently throws an ambiguous error because the Powershell command requires a backslash at the beginning of the path.

dagwieers replied:
@DepthDeluxe I understand, but I don't think we are doing anyone a favor to turn relative paths into absolute paths. We may be misleading them into thinking it is working differently than they are expecting. Better be strict about it and make things clear from that start.

mellena1 commented on lib/ansible/modules/windows/win_scheduled_task.ps1:
```diff
@@ -135,19 +152,25 @@ try {
             $trigger = New-ScheduledTaskTrigger -Once -At $time
         }
+        elseif ( $frequency -eq "interval"){
+            $trigger = New-ScheduledTaskTrigger -Once -RepetitionInterval (iex $("new-timespan -$interval_unit $interval")) -At $time -RepetitionDuration ([timespan]::MaxValue)
```
I've been messing around with this script a bit and I think I found a potential problem here. At least in my testing, having `-RepetitionDuration ([timespan]::MaxValue)` as a flag in windows server 2016 (probably windows 10 too) breaks the script with a formatting error when the command generates the xml. Not having the flag for windows server 2012 breaks it as well because that version requires a flag. My solution was:
```
If (([System.Environment]::OSVersion.Version).Major -eq 10){
    $trigger = New-ScheduledTaskTrigger -Once -RepetitionInterval (iex $("new-timespan -$interval_unit $interval")) -At $time
} Else {
    $trigger = New-ScheduledTaskTrigger -Once -RepetitionInterval (iex $("new-timespan -$interval_unit $interval")) -At $time -RepetitionDuration ([timespan]::MaxValue)
}
```
I'm sure there's a better way of doing that, but I'm not too familiar with powershell so that's what I found in a quick google search.

Excuse my ignorance if I missed something obvious here (maybe my powershell version in 2012 was too old or something).

jborean93 commented:
Thanks for your work @DepthDeluxe, I am going to close this PR as win_scheduled_task just went through a massive rewrite and now supports this functionality (see https://github.com/ansible/ansible/pull/28995). The original code had a few problems  and was only supported on newer hosts while now it works on all supported OS versions and you can now specify multiple actions and triggers.

Once again we appreciate the work you put into this and sorry we weren't able to get it merged in sooner.
